### PR TITLE
Use char code array in createTextWriter

### LINKED ISSUE #44036: Improve performance of the API returned by `createTextWriter`
# Suggestion

## 🔍 Search Terms

List of keywords you searched for before creating this issue. Write them down here so that others can find this suggestion more easily and help provide feedback.
- performance
- createTextWriter

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

Based on the performance wins in #44031 / #43987, there may be opportunities to improve the performance of concatenating lots of short strings in the `createTextWriter()` API.

## 📃 Motivating Example

Compiler performance. Currently print takes substantially longer than parse, so anything to improve print time without sacrificing code maintainability is a worthwhile endeavor.

## 💻 Use Cases

Compiler performance

### PR DESCRIPTION
**Pull request by dmichon-msft**
Inquiring minds wanted to know, based on the performance savings in #44031 , if the same trick would work in `createTextWriter`. This PR aims to answer that question.

Fixes #44036

### CODE REVIEW

**Comment by dmichon-msft** [comment]
@amcasey, @DanielRosenwasser : one of you mind kicking a perf test? Running the output on the TypeScript compiler `node ./built/local/tsc -p ./src/compiler/tsconfig.release.json --extendedDiagnostics` suggests it saves about 5%.

**Comment by dmichon-msft** [comment]
Thanks, @RyanCavanaugh

**Comment by dmichon-msft** [comment]
Hm, first glance at the benchmarks is not promising. I suspect whether or not this method is an improvement may depend heavily on how long the input strings to the writer are. Long comments that are preserved will result in redundant decode/encode work relative to the raw string concatenation. Might need a hybrid approach that bypasses the char code array for strings longer than some threshold.

File: src/compiler/utilities.ts [hunk 1]
```diff
@@ -3932,32 +3932,66 @@ namespace ts {
...
+
+            lineStart = linePos === totalChars;
         }
 
         function writeText(s: string) {
             if (s && s.length) {
                 if (lineStart) {
-                    s = getIndentString(indent) + s;
-                    lineStart = false;
+                    for (let i = 0; i < indent; i++) {
```
Comment (dmichon-msft) [inline src/compiler/utilities.ts:3989]: Needs to scale by `getIndentSize()`

**Comment by dmichon-msft** [comment]
Well, those results are fascinating. 5% improvement for TFS, 5% slower for Angular, about the same for Monaco and Compiler Unions.

File: src/compiler/utilities.ts [hunk 2]
```diff
@@ -3932,32 +3932,83 @@ namespace ts {
...
+        function appendRawLarge(text: string) {
+            flushBuffer();
+
+            const len = text.length;
+            for (let pos = 0; pos < len; pos++) {
+                const ch = text.charCodeAt(pos);
+                ++totalChars;
+                lastChar = ch;
+                // Ignore carriageReturn, since we mark the following lineFeed as the newline anyway
+                if (isLineBreak(ch) && ch !== CharacterCodes.carriageReturn) {
```
Comment (DanielRosenwasser) [inline src/compiler/utilities.ts:3989]: Technically carriage return can appear on its own and will be considered a newline.

**Comment by dmichon-msft** [comment]
@rbuckton , @weswigham , do either of you know what the key differences are between the Angular benchmark and the TFS benchmark that might explain the differences in results?

**Comment by DanielRosenwasser** [comment]
@typescript-bot perf test this

**Comment by dmichon-msft** [comment]
@weswigham So that means this might be higher overhead per file but scale better for large files?

**Comment by dmichon-msft** [comment]
This looks like a no go so far. I've explored altering `writePunctuation()`, `writeSpace()`, `writeTrailingSemicolon` to write the character codes directly (i.e. without having to decode them from a string first), and that didn't seem to help much if at all. The build is still slower against my reference codebase.

**Pull request closed without merging**